junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/HierarchicalTestExecutor.java

```diff
@@ -69,6 +70,7 @@ private void execute(TestDescriptor testDescriptor, C parentContext, ExecutionTr
 			// We call executionStarted first to comply with the contract of EngineExecutionListener
 			this.listener.executionStarted(testDescriptor);
 			this.listener.executionFinished(testDescriptor, TestExecutionResult.failed(throwable));
+			// node.cleanup(preparedContext); // Variable `preparedContext` might not have been initialized.
```

Comment:
If you initialize `null` to `preparedContext` you can call `cleanup()` when it's not `null`.

Reply:
"effective final" constraint is killed by that initialization. Will break the flow some lines deeper.